### DISCUSSION TITLE
Close : Print Icons View

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#175] [General]
Of course the goal is not to print a catalog of 1000 covers.
but to put forward with one or 2 pages of books (novelties; detective series; various theme Egypt; Gardening; BD Child to make discover our bottom with the reader: one passes has quoted treasure in the racks :-)

[A4-print.pdf](https://github.com/textbrowser/biblioteq/files/7844729/A4-print.pdf)

for vertical format :

center title :
 01 02 03 04 05
 06 07 08 09 10
 11 12 13 14 15
 16 17 18 19 20
 21 22 23 24 25

for horizontal format :

center title
01 02 03 04 05 06 07 08
09 10 11 12 13 14 15 16
17 18 19 20 21 22 23 24

1. a small problem of page-setting, it does not use the whole width 
2. to be able to put a title to know what we present
3. the saving directory of the pdf file should be /home/user by default (even if I like home/user/biblioteq :-) but for the users ...

test base : 
[2021-11-27_liste-47_achat_nouveautés.sqlite.zip](https://github.com/textbrowser/biblioteq/files/7844730/2021-11-27_liste-47_achat_nouveautes.sqlite.zip)

just an idea :  (many pdf's no info has by the producer (the software)
4 - extra options : to be able to fill the pdf fields (title ; subject ; author ; keyword ; creator)(you never know BiblioteQ might get the info automatically from the pdf) and replace Qt 5.12.8 with BiblioteQ for the creator :-)

**Comment by meteos77** [comment]
for my request it is very limited  no page with the 8000 images to put in the png
just a page with 24 or 25 thumbnails and a title but with a correct layout

process, we make the filter with sql SELECT ... with the accession_number it remains only 24 documents ... export to png horizontal format ... specify the title ... print to png file

or 25 documents export to png vertical format ... print to png file

**Comment by textbrowser** [comment]
This is completed and included in the release notes. The export and the print tasks perform their duties on the current view's contents. That is, the context is important and it's the person's responsibility to be aware of the context. :)

**Reply by meteos77** [reply]
Great feature
Suggestion without urgency: Wouldn't the icons mode view be better horizontally to display more images? (most screens are wider than high with the current view with my 1900 x 1000 resolution: I display 20 covers on my screen with the full screen.

**Comment by textbrowser** [comment]
Most is subjective. On a tablet, no. On a large resolution, it depends on the scene's viewable size. The number of columns would depend upon the size of the scene. Five columns was a reasonable estimate ten years ago. :D

**Comment by textbrowser** [comment]
BQ is slowly adapting to more options. So, a custom approach is again the best solution and not a solution the software applies because of its environment.

**Reply by meteos77** [reply]
the different software I follow regularly ... none of them add options continuously as you do with BiblioteQ... again, congratulations for this feat

**Reply by textbrowser** [reply]
More options, more silence. :)

**Reply by meteos77** [reply]
adding new options brings new comments and new ideas :-)

**Comment by textbrowser** [comment]
I don't like software which estimates its environment or pretends to believe that its defaults are best. I like software which is malleable and allows the person to think and decide.

**Comment by textbrowser** [comment]
And such software is expensive to make because options are like people.

**Comment by textbrowser** [comment]
New Other Options completed. Range is [5, 100]. Number of columns.

**Reply by meteos77** [reply]
Great option, it will allow to manage the display with the available space on his screen.

small problem with the PNG export, it always leaves a white space on the right ?
whatever the number of icons

**Comment by textbrowser** [comment]
The image's size is set to the size of the bounding rectangle of the scene. The bounding rectangle is the smallest rectangle which contains all of the items. It may or may not have so-called empty areas.

**Comment by meteos77** [comment]
example after a few tries
![01-Nouveautés-BD-Adultes-2022-06-10-colonnes-affichage-50-web](https://user-images.githubusercontent.com/56431039/160901566-2c71b7fe-1c02-480e-9fe3-944b7a22c653.png)

**Comment by textbrowser** [comment]
How many tries?

**Reply by meteos77** [reply]
about 20  - I always had spaces

**Reply by textbrowser** [reply]
Too many. What can improve it?

**Reply by meteos77** [reply]
to get the best result, you have to put the option
articles per page closest to the maximum number of thumbnails
if 49 documents -> set 50

on the other hand if 49 documents with 30 => a lot of white space
![1-30](https://user-images.githubusercontent.com/56431039/160904727-88a988b8-cb39-4c05-84b7-877496fed3ca.png)

**Reply by meteos77** [reply]
I don't know yet, I'm trying to understand the logic of the investment

**Comment by meteos77** [comment]
my test base
[Base_Travail-Export-PNG-BDA.sqlite.tar.gz](https://github.com/textbrowser/biblioteq/files/8383135/Base_Travail-Export-PNG-BDA.sqlite.tar.gz)

**Reply by meteos77** [reply]
for example
search type of binding : broché
result : 3 books

**Reply by meteos77** [reply]
if I take the original base with 296 documents
by default I have this result
![1](https://user-images.githubusercontent.com/56431039/160907200-2fc81fc0-dfb2-43ad-8e41-2f23be98ae3d.png)

**Reply by meteos77** [reply]
idea: maybe the layout takes the total number of the base and not the total number of the selection

**Reply by meteos77** [reply]
to get my first result
1 - copy of my complete database
2 - delete the documents I don't want to export in PNG
3 - setting the display closest to the number of documents (above) 49 documents so 50
4 - export the view to PNG

**Reply by meteos77** [reply]
my original database 
![1](https://user-images.githubusercontent.com/56431039/160908600-b4080bcb-6dba-41b5-81b8-485c8788f087.png)

**Reply by meteos77** [reply]
when you print the whole database it is well presented

**Comment by meteos77** [comment]
PNG export experience : report a little more precise

at startup I have the application in table mode (no switch to icon mode)
I select by Book Search and I choose a location

Test 1 to start the application in table mode (no switch to icon mode)
image 1 : selection de Classe 300 - Documentaites Sciences sociales => 5 Result(s)    **Export PNG OK**

image 2 : selection de Classe 600 - Documentaires Technologie ... => 39 Result(s)       **Export PNG OK**

image 3 : selection de BDA- Bande Déssinée Adultes => 49 Result(s)                            **Export PNG OK**

re-starting BQ   
image 4 : selection de BDA - Bande Déssinée Adultes => 49 Result(s)                           **Export PNG OK**
image 5 : selection de Classe 600 - Documentaires Technologie ... => 39 Result(s)          **Export PNG KO KO**
image 6 : selection de Classe 300 - Documentaites Sciences sociales => 5 Result(s)       **Export PNG KO KO**

[Rapport-Mode-Tableau.tar.gz](https://github.com/textbrowser/biblioteq/files/8404162/Rapport-Mode-Tableau.tar.gz)

**Comment by meteos77** [comment]
Test 2 at startup to have the application in icon mode (no switch to table mode)
image 7 : selection de Classe 300 - Documentaites Sciences sociales => 5 Result(s)    **Export PNG KO KO**
image 8 : selection de Classe 600 - Documentaires Technologie ... => 39 Result(s)     **Export PNG KO**
image 9 : selection de BDA- Bande Déssinée Adultes => 49 Result(s)                    **Export PNG KO**

re-starting de BQ
image 10 : selection de BDA - Bande Déssinée Adultes => 49 Result(s)                   **Export PNG KO**
image 11 : selection de Classe 600 - Documentaires Technologie ... => 39 Result(s)     **Export PNG KO**
image 12 : selection de Classe 300 - Documentaites Sciences sociales => 5 Result(s)    **Export PNG KO**

[Rapport-Mode-Icons.tar.gz](https://github.com/textbrowser/biblioteq/files/8404167/Rapport-Mode-Icons.tar.gz)

**Comment by meteos77** [comment]
my conclusion :

In table mode if the number of thumbnails to export is always higher than the one of the PNG export before everything is fine.
on the other hand if it is lower the white spaces appear

In icons mode, as the first display is all the icons or the total of the base, BQ exports only when we take the totality
image 13-ALL
[13-ALL.png.tar.gz](https://github.com/textbrowser/biblioteq/files/8404178/13-ALL.png.tar.gz)

**Comment by textbrowser** [comment]
Would setting each image to a fixed size during the PNG export create a nicer result?

**Comment by textbrowser** [comment]
To a uniform size, rather.

**Comment by textbrowser** [comment]
The white space is from the graphics scene because an export as PNG is essentially a copy of the graphics scene. Here's an export of the sample database.

![test](https://user-images.githubusercontent.com/10701156/161539375-90b6ba09-5117-4f13-8a48-9ef15556ba31.png)

**Reply by meteos77** [reply]
For me the layout is perfect as it looks now (when it works well)
the "Icons View Column Count" setting also works fine

for the table view, it's just the problem when the number of thumbnails is lower than the previous request that is a problem. probably a variable that doesn't reset to zero like when you start BiblioteQ

for the export by the icons view it is necessarily the totality of the documents (just specify it in the documentation)

**Reply by meteos77** [reply]
after this export, if you select just the documents concerning cryptography, the layout should put blanks on the first line

**Reply by textbrowser** [reply]
> For me the layout is perfect as it looks now (when it works well) the "Icons View Column Count" setting also works fine
> 
> for the table view, it's just the problem when the number of thumbnails is lower than the previous request that is a problem. probably a variable that doesn't reset to zero like when you start BiblioteQ
> 
> for the export by the icons view it is necessarily the totality of the documents (just specify it in the documentation)

Can you describe the table problem?

**Reply by meteos77** [reply]
this is the message above in the discussion

if the number of thumbnails is 10 during the 1st export it displays correctly the 5 on 2 lines

if at the next export BQ must display 20 thumbnails or 5 on 4 lines everything is OK
on the other hand if with the next export BQ must post 5 labels then the 5 labels are well on a line but all tightened, one would say that it keeps the calculation which it carried out at the time of the preceding export.

**Reply by textbrowser** [reply]
Interesting. Should this be investigated?

**Reply by meteos77** [reply]
With the little programming knowledge I have I would say that line 148 of biblioteq_d.cc
(ui.graphicsView->scene()->itemsBoundingRect().size().toSize(),

the toSize variable is the problem 

I let the specialist do it to see if I was on the right track :-)

**Comment by textbrowser** [comment]
Perhaps. However, this size depends upon the bounding rectangle of the present items and not future or past items.

**Reply by meteos77** [reply]
I may be following a wrong track :-)
it works well when BiblioteQ is started, whatever the number of thumbnails is **in table mode**

**Reply by meteos77** [reply]
> Perhaps. However, this size depends upon the bounding rectangle of the present items and not future or past items.

Does that mean he doesn't recalculate?

**Reply by textbrowser** [reply]
See below tests. The scene's bounding rectangle is correct.

**Comment by textbrowser** [comment]
I see that too.

**Comment by textbrowser** [comment]
First export (5 items per page): QSizeF(683, 188).
Set 10 items per page, second export: QSizeF(683, 388).
Set 5 items per page, third export: QSizeF(683, 188).

Third export is faulty. Sizes are correct for all tests.

**Reply by textbrowser** [reply]
So it's not the bounding rectangle. At least according to those sizes.

**Comment by textbrowser** [comment]
I tried invalidating the scene after it's cleared. Does not help.

**Comment by textbrowser** [comment]
Let's see if we can disable caching.

**Reply by meteos77** [reply]
what is the difference between a brand new BQ at startup and once it has rendered a PNG export?

**Comment by textbrowser** [comment]
Sometime before a scene is populated, it's cleared. It's also resized to fit the predicted number of images.

**Comment by textbrowser** [comment]
As you see the correct items and correct sizes, the export should be correct.

**Comment by textbrowser** [comment]
The difference is complicated. :P It's a new process. However, the scene is essentially reset before it's populated.

**Comment by textbrowser** [comment]
Disabling the cache did not help.

**Reply by meteos77** [reply]
I know why I am not a programmer :-)

**Comment by textbrowser** [comment]
Neither the cache nor the invalidation helps. Sizes are correct.

**Reply by meteos77** [reply]
qt problem?

**Reply by meteos77** [reply]
for my information
everything happens in void biblioteq::slotExportAsPNG(void) of the file "biblioteq/Source/biblioteq_d.cc" and in the 30 lines ?

**Comment by meteos77** [comment]
the variable that you enter in the options to choose the number of columns?

**Comment by textbrowser** [comment]
Try now.

**Reply by meteos77** [reply]
you are the magician ... thank you very much

**Comment by textbrowser** [comment]
The render() function render's the scene's rectangle. One has to ask it to render the scene's bounding rectangle instead.

**Comment by meteos77** [comment]
this function is a 5* :-)
